### ISSUE TITLE
access_token resource CORS preflight incorrectly requires auth

### ISSUE THREAD

**Issue by jeff-h** [RESTful-Drupal/restful#844]
Phew, what a great issue title :dancer: 

Explanation: when attempting a `GET` to the `access_token` resource from a different domain to the RESTful Drupal server e.g.

``` js
jQuery.ajax({
    url: "http://drupalserver.com/api/login-token",
    type: "GET",
    headers: {
        "Authorization": "Basic bhcRlOn*sanitised*3N3bWFWJk",
        "Content-Type": "application/octet-stream",
    },
})
```

...jQuery correctly sends a preflight `OPTIONS` request first. However, it doesn't send the auth credentials with the `OPTIONS` request.

This is correct behaviour as per [the W3 spec for CORS preflight requests](https://fetch.spec.whatwg.org/#cors-preflight-fetch) which states that user credentials should be excluded from CORS preflight OPTIONS requests (1).

_However_, RESTful's access_token resource is returning a `401` for this preflight `OPTIONS` request.

I'm not sure if RESTful's architecture allows auth to be optional depending on the request type. My work is currently a cycle between lots of Drupal/RESTful then lots of Marionette and currently my head's deep in JS so sorry if this is something I should have fixed easily.
1. http://stackoverflow.com/a/15734032/2778502

**Comment by jeff-h** [comment]
Just tried my JS app in Chrome and it won't even attempt the `GET` because the preflight is returning `401`. Seems Safari is willing to do the `GET` anyway.

Basically this issue breaks my app in Chrome.

**Comment by jeff-h** [comment]
It's quite possible I'm way off but after thinking this through, I fear there's an underlying problem, which is that RESTful should not use `OPTIONS` requests for resource discovery.

I would have expected an authentication-protected resource to only provide its discovery information to appropriately authenticated users, but as per the W3 spec, `OPTIONS` requests shouldn't ever be authenticated.

If you take the stance that discovery should not be subject to authentication, then RESTful should never return `401` for any `OPTIONS` requests (unfortunately it does right now).

Another reason not to use `OPTIONS` for discovery is that I think `OPTIONS` calls should be as lightweight as possible, given that they are most often the first of a double server hit. Drupal is bootstrapping twice for every request using CORS, right?

I went looking around the internet for discussion about resource discovery.

##### Google:

I noticed Google use a different URL for discovery, meaning discovery requests to their API are `GET` requests, with a `/discovery/` URL prefix. e.g.

`GET https://www.googleapis.com/discovery/v1/apis/urlshortener/v1/rest`

(see https://developers.google.com/apis-explorer/#p/discovery/v1/discovery.apis.getRest?api=urlshortener&version=v1&_h=1& for an example)

##### Blog post by Mark Nottingham (W3C TAG):

Kind of old (2012), but on-topic: https://www.mnot.net/blog/2012/10/29/NO_OPTIONS

**Comment by fpalluel** [comment]
Hello,
I faced this problem with Firefox too and had to "dirty patch" the authentication manager to accept OPTIONS request without credentials. I had to modify settings.php too, to add necessary response headers...

**Comment by e0ipso** [comment]
Yikes!

This issue description is very thorough. As you point out:

> I fear there's an underlying problem, which is that RESTful should not use `OPTIONS` requests for resource discovery.

Undoing this at this moment would be an API break requiring me to spawn `7.x-3.x`, which I would like to avoid.

A _quick hack_ would be:
1. Exclude `OPTIONS` from Authentication checking (potentially hard).
2. If serving `OPTIONS` and authentication is enabled:
   
   2.1 If authentication credentials are valid: respond like we are doing now.
   2.2 If authentication credentials are not valid: respond like we are doing now but with a blank body.

**Comment by jeff-h** [comment]
Thanks for giving this your brain time — I reckon this would be the kind of issue you'd wish would go away by itself :)

The RESTful users who will be affected by this are:
- those running a JavaScript REST client inside a browser from a different domain to the Drupal instance, and are thus using CORS

and who
- need to use authentication

I'm not sure how many people are right now using RESTful to provide REST services to a JavaScript web app that also requires auth, but I think this may be a small enough subset that those who need a fix will find this discussion and understand what RESTful is offering.

Personally, I'd love it if the 3.x branch (when it comes) does ditch `OPTIONS` for resource discovery, but if your suggested option 2.2 above is doable that would fix the issue for now.

_However_, I wanted to also add that yesterday I gave [xdomain](https://github.com/jpillora/xdomain) a shot and it really works! **CORS is basically incredibly stupid** in the first place, and xdomain is a clever little hack that avoids it completely. It doubles my app's responsiveness into the bargain, so I definitely recommend this as a solution.

**Comment by e0ipso** [comment]
Awesome! I did not know about [xdomain](https://github.com/jpillora/xdomain). I agree that the affected users are small subset.

Provided that there is an alternative, this is probably going to remain as a _known bug_ that I will not fix. That makes me sad, but I can use the free time.

**Comment by jeff-h** [comment]
Kick back, and send some nice tweets :D

**Comment by e0ipso** [comment]
You're twitter handle is https://twitter.com/marmaladesoul after all? I was not sure about it.

**Comment by jeff-h** [comment]
It is! :dancers:

**Comment by jeff-h** [comment]
(I put some notes into the wiki at https://github.com/RESTful-Drupal/restful/wiki/OPTIONS-requests#a-warning-regarding-cors-and-authenticated-resources)

**Comment by omarlopesino** [comment]
@e0ipso In order to exclude authentication checking, I made a workaround to _\Drupal\restful\Plugin\resource\Resource_ class. At the end of _initAuthenticationManager_ method, I added this code:

```
if ($_SERVER['REQUEST_METHOD'] == 'OPTIONS') {
  $this->authenticationManager->setIsOptional(TRUE);
}
```

So, if method is OPTIONS authentication will be set allways as optional. Adding this, it worked for me in the requested resource (login), receiving a 200. 

Can this workaround completely works to skip authentication in OPTIONS method, or there can be problems I'm not aware of?
